library/std/src/io/error/repr_bitpacked.rs

```diff
@@ -0,0 +1,346 @@
+//!   meaning its tagged and untagged representation are equivalent.
+//!
+//!   This means we can skip tagging it, which is necessary as this variant can
+//!   be constructed from a `const fn`, which probably cannot tag pointers (or
+//!   at least it would be difficult.
+//!
+//! - **Tag 0b01**: The other pointer variant holds the data for
+//!   `ErrorData::Custom` and the remaining 62 bits are used to store a
+//!   `Box<Custom>`. `Custom` also has alignment >= 4, so the bottom two bits
+//!   are free to use for the tag.
+//!
+//!   The only important thing to note is that `ptr::add` and `ptr::sub` are
+//!   used to tag the pointer, rather than bitwise operations. This should
+//!   preserve the pointer's provenance, which would otherwise be lost.
+//!
+//! - **Tag 0b10**: Holds the data for `ErrorData::Os(i32)`. We store the `i32`
+//!   in the pointers most significant 32 bits, and don't use the bits `2..32`
+//!   for anything. Using the top 32 bits is just to let us easily recover the
+//!   `i32` code with the correct sign.
+//!
+//! - **Tag 0b11**: Holds the data for `ErrorData::Simple(ErrorKind)`. This
+//!   stores the `ErrorKind` in the top 32 bits as well, although it doesn't
+//!   occupy nearly that many. Most of the bits are unused here, but it's not
+//!   like we need them for anything else yet.
+//!
+//! # Use of `NonNull<()>`
+//!
+//! Everything is stored in a `NonNull<()>`, which is odd, but actually serves a
+//! purpose.
+//!
+//! Conceptually you might think of this more like:
+//!
+//! ```ignore (exposition-only)
+//! union Repr {
+//!     // holds integer (Simple/Os) variants, and
+//!     // provides access to the tag bits.
+//!     bits: NonZeroU64,
+//!     // Tag is 0, so this is stored untagged.
+//!     msg: &'static SimpleMessage,
+//!     // Tagged (offset) `Box<Custom>` pointer.
+//!     tagged_custom: NonNull<()>,
+//! }
+//! ```
+//!
+//! But there are a few problems with this:
+//!
+//! 1. Union access is equivalent to a transmute, so this representation would
+//!    require we transmute between integers and pointers in at least one
+//!    direction, which may be UB (and even if not, it is likely harder for a
+//!    compiler to reason about than explicit ptr->int operations).
+//!
+//! 2. Even if all fields of a union have a niche, the union itself doesn't,
+//!    although this may change in the future. This would make things like
+//!    `io::Result<()>` and `io::Result<usize>` larger, which defeats part of
+//!    the motivation of this bitpacking.
+//!
+//! Storing everything in a `NonZeroUsize` (or some other integer) would be a
+//! bit more traditional for pointer tagging, but it would lose provenance
+//! information, couldn't be constructed from a `const fn`, and would probably
+//! run into other issues as well.
+//!
+//! The `NonNull<()>` seems like the only alternative, even if it's fairly odd
+//! to use a pointer type to store something that may hold an integer, some of
+//! the time.
```

Comment:
FWIW, I think it would be even more odd to use an integer type to store something that may hold a pointer, some of the time. ;)  But I understand that is the "more traditional" thing to do.

I fully agree with your choice here.

Reply:
I think the logic is to only use a pointer if it is always a pointer, and to use an integer if it ever is instead used to store arbitrary integer data.

I'm pretty sure the reason for this is that the C and C++ standards get very cagey about promising int->ptr->int will round-trip (they don't promise it, even if the int is big enough), but they do guarantee that ptr->int->ptr will for a sufficiently sized int.

OTOH, Rust's approach actually makes more sense to me when I think about it — there are cases where the compiler really absolutely needs to know something is a pointer just to correctly compile the code (for example, if it's a constant, and it points something which may be relocated).

Reply:
> I'm pretty sure the reason for this is that the C and C++ standards get very cagey about promising int->ptr->int will round-trip (they don't promise it, even if the int is big enough), but they do guarantee that ptr->int->ptr will for a sufficiently sized int.

The irony is that ptr->int->ptr in C/C++ will almost certainly *not* roundtrip losslessly (so, yes it does a roundtrip, but information is lost). That is demonstrated by [this example](https://www.ralfj.de/blog/2020/12/14/provenance.html). In contrast, int->ptr->int will roundtrip without loss at least in LLVM. (But in C, int->ptr is UB if this is not a valid address...)

An alternative would be to use a `union` for this:
```rust
union PtrOrInt {
  ptr: NonNull<()>,
  int: NonZeroUsize,
}
```
You can always access the `ptr` field to check the tag (since int->ptr transmute is fine, unlike ptr->int transmute which is [problematic](https://github.com/rust-lang/unsafe-code-guidelines/issues/286)).

But that is actually fairly close to just using `NonNull` like you do, the only difference being what happens when turning an integer value into the tagged representation.

Reply:
> there are cases where the compiler really absolutely needs to know something is a pointer just to correctly compile the code (for example, if it's a constant, and it points something which may be relocated).

This is just saying that provenance exists for real. :)  At compile-time, there's no denying this.